# **Tarea 1 - CC6205 Natural Language Processing 📚**

**Integrantes:**
Ismael Fernandez
Yerko Garrido 

**Fecha límite de entrega 📆:** Martes 4 de Abril.

**Tiempo estimado de dedicación:** 




` ` 


Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en ***Information Retrieval (IR)*** y ***Vector Space Models***. Si aún no han visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP. 

` ` 




**Instrucciones:**
- La tarea se realiza en grupos de **máximo** 2 personas. Puede ser invidivual pero no es recomendable.
- La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
- El formato de entrega es este mismo Jupyter Notebook.
- Al momento de la revisión tu código será ejecutado. Por favor verifica que tu entrega no tenga errores de compilación. 
- Está **PROHIBIDO** usar cualquier librería que implemente los algoritmos pedidos (Spacy, scikit, etc). Sólo se podrán utilizar las librerías importadas al inicio de la sección de práctica.
- En el horario de auxiliar pueden realizar consultas acerca de la tarea a través del canal de Discord del curso. 



Ahora sí, empecemos! 😄🚀

` ` 

**Referencias:**


Slides:
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    

Videos: 

- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte I](https://www.youtube.com/watch?v=HEKTNOttGvU)
- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte II](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 1](https://youtu.be/FXIVClF370w)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 2](https://youtu.be/f8nG1EMmPZk)

` ` 



## **1 - Preguntas teóricas 📕 (2 puntos).** ##









Las siguientes celdas contienen preguntas acerca del contenido visto en clases y en el material del curso.  Contestar cada pregunta en su celda correspondiente y **no extenderse más de 100 palabras** . 🙏
` ` 

**Pregunta 1 (0.2 puntos): ¿Por qué el análisis del lenguaje humano es una tarea compleja? Mencione dos razones según lo visto en clases. Debe citar la slide donde cree que está la respuesta.**




``
* El lenguaje humano es muy ambiguo: en la diapo n°4 de la introducción se presenta el ejemplo de la oración de comer pizza, que a pesar de tener la misma estructura gramatical, se va modificando el sentido del verbo en función de la última palabra. Es decir de comer pizza con amigos, comer pizza con olivas o comer pizza con tenedor. 

* El Lenguaje humano esta siempre evolucionando: En la diapo n°4 de la introuducción, se presenta que el lenguaje esta en constante evolución y que el idioma español actual, es distinto al idioma español de nuestros antepasados.
` ` 

**Pregunta 2 (0.4 puntos): ¿Cuáles son las diferencias entre usar Deep learning y Machine Learning clásico (empirismo) para un problema de NLP? Ejemplifique con alguna task.** Puede utilizar ChatGPT (debe indicarlo) para generar la respuesta y luego debe indicar si la respuesta entregada por ChatGPT es correcta o no. Mencione por qué según lo visto en clases.


` ` 
El deep learning es capaz de extraer las relaciones a partir de los datos, en cambio el Machine Learning necesitan de texto etiquetado. Etiquetar es costoso y puede ser sesgado. Por ejemplo en clasificación de texto si se realiza con entrenamiento con documentos con información deportiva no se comportara bien cuando se intente clasificar documentos de politica.

` ` 

 **Pregunta 3 (0.4 puntos) Según las primeras clases, ¿Qué método clásico nos permite rankear las similitudes existentes entre documentos?, ¿Cómo son las representaciones que genera y problemas que podrían experimentar estas soluciones simples?** 

` ` 
El metodo clásico es el de indice invertido. Genera una representación del tipo diccionario. Mapea los terminos del vocabulario dentro del texto del siguiente modo: < term > -> < docId >*
` ` 

**Pregunta 4 (0.4 puntos) Usted se encuentra realizando un modelo de clasificación de sentimientos con texto, su jefe le señala que debe eliminar las palabras mas comunes para obtener una mejor clasificación. ¿Qué palabras le señala que elimine su jefe?, ¿es acaso esto una buena idea?.** Puede utilizar ChatGPT (debe indicarlo) para generar la respuesta y luego debe indicar si la respuesta entregada por ChatGPT es correcta o no. Mencione por qué según lo visto en clases.

` ` 
El jefe se refiere a que se deben eliminar las stopwords. Depende de el modelo. Puede ser que dentro de las stop words que son articulos, sustantivos, etc. En analisis de sentiemientos, puede generar problemas al borrar palabras que permiten la negación.
` ` 

**Pregunta 5 (0.6 puntos): Acorde al paper [A Vector Space Model for automatic indexing](https://dl.acm.org/doi/pdf/10.1145/361219.361220) un documento, $D_i$, puede ser definido formalmente como una tupla de términos, $(d_{i1}, d_{i2}, ..., d_{in})$, donde $d_{ij}$ representa el peso del j-esimo término. En clase vieron algunas formas medir los pesos de estos términos. Mencione cuales fueron y sus ventajas y desventajas (pueden responder en máximo 150 palabras).** 

` ` 
Respuesta aquí
` ` 

## **2 - Preguntas prácticas 💻 (4 puntos).** ##

Esta segunda sección incluye ejercicios de programación 🤙. Leer atentamente las instrucciones entregadas a continuación para facilitar el proceso de revisión de sus trabajos.


` ` 

**Instrucciones:**



- Escribe tu código entre las lineas de comentarios **### Aquí inicia tu código ###** y **### Aquí termina tu código ###**.
- Cuando el ejercicio incluya un bloque llamado ***Test***, comprueba que el resultado de la ejecución coincida con el resultado esperado.
- Recuerde siempre mantener buenas prácticas de código.
- Está permitido sólo utilizar las librerías importadas antes del Ejercicio 1.
- **Recordar** que: *Documento = Oración. Dataset = Corpus. Vocabulario = Tokens*.
- El **orden de los resultados** pueden variar dependiendo de su máquina, pero los valores de los resultados son los mismos.



**Ejemplo:** Implemente una función **`hello_world()`** que imprima en pantalla `"Hello World"`. 

In [ ]:
def hello_world():
  ### Aquí inicia tu código ###
  print("Hello World")
  ### Aquí termina tu código ###

***Test:***

In [ ]:
hello_world()

Hello World


***Resultado esperado***: 
<table>
    <tr> 
        <td> Hello World </td> 
    </tr>
</table> 

``
``




Estas son las librerías permitidas. Si quieren utilizar alguna librería adicional, pueden realizar la consulta a través de Discord. 

In [1]:
import codecs
import re
import numpy as np
import pandas as pd

En caso de desarrollar la tarea desde colab, utilizar el siguiente código para cargar los archivos desde el drive:

In [ ]:
#try:
#    from google.colab import drive#

#    drive.mount("/content/drive")
#    path = 'Colab Notebooks/NLP/Tarea_1/marcianeke.txt'
#except: 
#    print('Ignorando conexión drive-colab')

Mounted at /content/drive


` ` 

**Ejercicio 1 - *Tokenización* (0.5 puntos).** 
` `  
` ` 

En el primer ejercicio veremos la dificultad 😨 de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo. 

El archivo adjunto al enunciado de la tarea contiene la letra de una canción del marcianeke 👽. Utilice este texto para realizar su primera tokenización y ver qué tan bien funciona su función. 

Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook.

In [4]:
path='marcianeke.txt'

In [5]:
text = codecs.open(path, 'r', 'UTF-8').read()
print(text)

Brr
Marcianeke
Vamo' a estar con Pailita
Dimelo má
Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar (brr)
Tussi, keta quiere' mezclar
Dimelo má

Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar
Tussi, keta pura quiere' mezclar
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má

Esperame que ahora entro yo
Y lo que pide yo lo traje
No visto de traje
Puro corte calle, no de maquillaje
Pronto coronamos y nos vamo' de viaje
Tanto hit que hago que lo' culo bajen
Ella se va de shopping
Sale positivo si se hace el doping
Baila twerk con un poco de popping
Los fardos en el botín

Si quieren letra llamen pa' mi booking
Generando, sigo en la mía lowkey
Cooking en el estudio con tu woman
Tanto whisky, pisco que hasta lo' vecinos toman
Si se tiran pa' aca puede que la arena coman
Ja, en el chanteo titulado sin diploma
Di-dimelo má
Di-di

Implementen una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Son libres de elegir la forma de tokenizar mientras no utilicen librerías con tokenizadores ya implementados. Pueden utilizar la librería **re** importada para trabajar símbolos.


Ejemplo de uso:

`get_tokens('Este es un ejemplo de prueba.')` 

Nos entregaría:

`['Este', 'es', 'un', 'ejemplo', 'de', 'prueba', '.']`

In [8]:
from posixpath import split
def get_tokens(text):
  ### Inicio del código ###
  tokens_=[]
  tokens=re.sub(r'\r\n{1,}',' ',text)
  tokens=tokens.split(' ')

  for token in tokens:
    for i in re.split(r'\W',token):
      if len(i)>=1:
        tokens_+=[i]
    tokens_+= re.findall(r'\W',token)

  ### Fin del código ###
  return tokens_

In [7]:
tokens = get_tokens(text)
tokens

['Brr',
 'Marcianeke',
 'Vamo',
 "'",
 'a',
 'estar',
 'con',
 'Pailita',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 'ah',
 '(',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 'rata',
 'ta',
 '(',
 '-',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 'brr',
 '(',
 ')',
 'Tussi',
 ',',
 'keta',
 'quiere',
 "'",
 'mezclar',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 'ah',
 '(',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 'rata',
 'ta',
 '(',
 '-',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 'Tussi',
 ',',
 'keta',
 'pura',
 'quiere',
 "'",
 'mezclar',
 'Di',
 'dimelo',
 '-',
 'má',
 'Di',
 'dimelo',
 '-',
 'má',
 'Di',
 'dimelo',
 '-',
 'má',
 'Di',
 'dimelo',
 '-',
 'má',
 'Di',
 'dimelo',
 '-',
 'má',
 'Di',
 'dimelo',
 '-',
 'má',
 'Di',
 'dimelo',
 '-',
 'má',
 'Esperame',
 'que',
 'ahora',
 'entro',
 'yo',
 'Y',
 'lo',
 'que',

**Describa cuáles fueron sus supuestos para realizar la tokenización y compare sus tokens con los entregados por la librería nltk en el bloque de código de más abajo.**

Supuestos aqui


In [9]:
from nltk.tokenize import wordpunct_tokenize 
nltk_tokens = wordpunct_tokenize(text)
nltk_tokens

['Brr',
 'Marcianeke',
 'Vamo',
 "'",
 'a',
 'estar',
 'con',
 'Pailita',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 '(',
 'brr',
 ')',
 'Tussi',
 ',',
 'keta',
 'quiere',
 "'",
 'mezclar',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 'Tussi',
 ',',
 'keta',
 'pura',
 'quiere',
 "'",
 'mezclar',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Esperame',
 'que',
 'ahora',
 'entro',
 'yo',
 'Y',
 'lo',
 'que',

` `  
` ` 

**Ejercicio 2 - *Stopwords y Stemming* (1 punto).** 
` `  
` ` 

Considere el siguiente corpus:


In [10]:
dataset = ["I like human languages", "I like programming languages", "Spanish is my favorite language"]

Diseñe una función **`get_vocab()`** que extraiga los tokens de este corpus solamente tokenizando. Puede utilizar la función del Ejercicio 1.

In [36]:
def get_vocab(dataset):
  
  ### Aquí inicia tu código ###
  vocab=[]
  for data in dataset:
    vocab_ = get_tokens(data)
    #apply lowercase:
    vocab_ = [word.lower() for word in vocab_]
    vocab+=vocab_
  return list(set(vocab))
  ### Aquí termina tu código ###

***Test:***

In [37]:
vocab = get_vocab(dataset)
vocab

['i',
 'languages',
 'favorite',
 'spanish',
 'programming',
 'my',
 'is',
 'language',
 'like',
 'human']

``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td>['favorite',
 'Spanish',
 'language',
 'I',
 'like',
 'programming',
 'languages',
 'my',
 'human',
 'is'] </td> 
    </tr>
</table> 

``
``

Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario actualizado. 

    Explique sus reglas aquí:


#### Remover StopWords:

In [38]:
#Diseñamos una lista de stopwords especifico para este dataset:
stopwords = ['is', 'my', 'i']



In [41]:
#check if a word is a stopword:
if 'my' in stop_words:
  print('is a stopword')

is a stopword


In [27]:
#download a stopwords list
import nltk
#nltk.download('stopwords')
from nltk.corpus import stopwords
stop_words = stopwords.words('english')
#print only the first 10 stopwords
stop_words


['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

is is a stopword


In [17]:
#function to remove stopwords
def remove_stopwords(tokens, stopwords):

    tokens_=[]
    for token in tokens:
        if token not in stopwords:
            tokens_.append(token)
    return tokens_


In [19]:
# remove stopwords from the tokens
vocab_sin_SW = remove_stopwords(vocab, stop_words)
vocab_sin_SW

['I',
 'languages',
 'favorite',
 'Spanish',
 'programming',
 'my',
 'is',
 'language',
 'like',
 'human']

#### Aplicar Steaming

In [ ]:
#funcion que aplica stemming a una lista de tokens:


In [ ]:
def pre_processing(vocab):

  pass

    

In [ ]:
vocab = pre_processing(vocab)
vocab


**Ejercicio 3 - *Bag of Words* 🐶🐈(0.5 puntos).** 
 


Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:

**disclaimer: El orden de los resultados pueden variar**

In [ ]:
d0 = 'El perro se come la comida y después se duerme'
d1 = 'El perro se despierta y después empieza a ladrar'
d2 = 'El perro ladra y después se come la comida'
d3 = 'El gato se come la comida y después se duerme'
d4 = 'El gato se despierta y después empieza a maullar'
d5 = 'El gato maulla y después se come la comida'
dataset = [d0, d1, d2, d3, d4, d5]

El objetivo de este ejercicio es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica TF-IDF. 

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es el **Bag of Words**, método mediante el cuál se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que reciba como input un arreglo de documentos y devuelva un pandas dataframe con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representa las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el bow de un determinado documento.


Por ejemplo para el siguiente dataset: 

```
dataset = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |



In [ ]:
def bag_of_words(dataset):
  ### Aquí inicia tu código ### 
  
  column=[]
  for i in dataset:
    column+=get_tokens(i)
  column=['docId']+list(set(column)) 

  df=pd.DataFrame()
  for i,j in enumerate(dataset):
    token=get_tokens(j)
    data={'docId':[f'd{i}']*len(token),'token':token, 'valor':[1]*len(token)}
    df_=pd.DataFrame(data=data)
    df=pd.concat([df,df_], axis=0, ignore_index=False)
  df=pd.pivot_table(df,values='valor', index='docId', columns='token',aggfunc='sum',fill_value=0).reset_index()
  df=df.fillna(0)
  df=pd.DataFrame(data=df[column].values.tolist(),columns=column)
  df.columns=column
  dfi=df.set_index('docId')
  del df
  return dfi


  ### Aquí termina tu código ###

***Test:***

In [ ]:
dataset_bow = bag_of_words(dataset)
dataset_bow

a  la  come  y  despierta  se  empieza  gato  El  maullar  maulla  \
docId                                                                      
d0     0   1     1  1          0   2        0     0   1        0       0   
d1     1   0     0  1          1   1        1     0   1        0       0   
d2     0   1     1  1          0   1        0     0   1        0       0   
d3     0   1     1  1          0   2        0     1   1        0       0   
d4     1   0     0  1          1   1        1     1   1        1       0   
d5     0   1     1  1          0   1        0     1   1        0       1   

       comida  duerme  ladrar  después  ladra  perro  
docId                                                 
d0          1       1       0        1      0      1  
d1          0       0       1        1      0      1  
d2          1       0       0        1      1      1  
d3          1       1       0        1      0      0  
d4          0       0       0        1      0      0  
d5          1       0       0        1      0      0

``
``

***Resultado esperado***: 

|    | El | perro | se | come | la | comida | y | después | duerme | despierta | empieza | a | ladrar | ladra | gato | maullar | maulla |
|----|---:|------:|---:|-----:|---:|-------:|--:|--------:|-------:|----------:|--------:|--:|-------:|------:|-----:|--------:|-------:|
| d0 |  1 |     1 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    0 |       0 |      0 |
| d1 |  1 |     1 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      1 |     0 |    0 |       0 |      0 |
| d2 |  1 |     1 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     1 |    0 |       0 |      0 |
| d3 |  1 |     0 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      0 |
| d4 |  1 |     0 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      0 |     0 |    1 |       1 |      0 |
| d5 |  1 |     0 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      1 |

``
``

` `  
` ` 

**Ejercicio 4 - *Calcular TF* (0.5 puntos):** Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia ${max_i({\text{tf}_{i,j})}}$, donde
i corresponde al índice de las filas (bow) y j al de las columnas (palabras). Es decir, dividir cada bow en la cantidad de veces de la palabra que aparezca más veces en ese vector. 


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{max_i({\text{tf}_{i,j})}}$$

In [ ]:
def calc_tf(dataset_bow):
    ### Aquí inicia tu código ###
    df=pd.DataFrame()
    for i,index in enumerate(dataset_bow.index):
      df_= dataset_bow.iloc[i, :]/np.max(dataset_bow.iloc[i, :].values.tolist()) 
      df=pd.concat([df,df_],axis=1)
    
    return df.T
    ### Aquí termina tu código ###

***Test:***

In [ ]:
tf = calc_tf(dataset_bow)
tf

a   la  come    y  despierta   se  empieza  gato   El  maullar  maulla  \
d0  0.0  0.5   0.5  0.5        0.0  1.0      0.0   0.0  0.5      0.0     0.0   
d1  1.0  0.0   0.0  1.0        1.0  1.0      1.0   0.0  1.0      0.0     0.0   
d2  0.0  1.0   1.0  1.0        0.0  1.0      0.0   0.0  1.0      0.0     0.0   
d3  0.0  0.5   0.5  0.5        0.0  1.0      0.0   0.5  0.5      0.0     0.0   
d4  1.0  0.0   0.0  1.0        1.0  1.0      1.0   1.0  1.0      1.0     0.0   
d5  0.0  1.0   1.0  1.0        0.0  1.0      0.0   1.0  1.0      0.0     1.0   

    comida  duerme  ladrar  después  ladra  perro  
d0     0.5     0.5     0.0      0.5    0.0    0.5  
d1     0.0     0.0     1.0      1.0    0.0    1.0  
d2     1.0     0.0     0.0      1.0    1.0    1.0  
d3     0.5     0.5     0.0      0.5    0.0    0.0  
d4     0.0     0.0     0.0      1.0    0.0    0.0  
d5     1.0     0.0     0.0      1.0    0.0    0.0

``
``

***Resultado esperado***: 

|    |  El | perro |  se | come |  la | comida |   y | después | duerme | despierta | empieza |   a | ladrar | ladra | gato | maullar | maulla |
|----|----:|------:|----:|-----:|----:|-------:|----:|--------:|-------:|----------:|--------:|----:|-------:|------:|-----:|--------:|-------:|
| d0 | 0.5 |   0.5 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d1 | 1.0 |   1.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    1.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d2 | 1.0 |   1.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   1.0 |  0.0 |     0.0 |    0.0 |
| d3 | 0.5 |   0.0 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.5 |     0.0 |    0.0 |
| d4 | 1.0 |   0.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    0.0 |   0.0 |  1.0 |     1.0 |    0.0 |
| d5 | 1.0 |   0.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  1.0 |     0.0 |    1.0 |

``
``

` `  
` ` 

**Ejercicio 5 - *Calcular IDF* (0.5 punto):**


Implementar `calc_idf(dataset_bow)`. Este debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el calculo de cada idf por palabra.

Recordar que $idf_{t_i} = log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ Número de documentos que contienen la palabra $t_i$ 

In [ ]:
def calc_idf(dataset_bow):
    ### Aquí inicia tu código ###
    N=dataset_bow.shape[0]    
    dataset_bow=dataset_bow.T
    palabras=dataset_bow.shape[0]
    diccionario={}

    for i in range(0,palabras):
      token=dataset_bow.reset_index().iloc[i,0]
      ni=np.sum(dataset_bow.iloc[i,0:].values.tolist())
      idf=np.log10(N/ni)
      diccionario[f'{token}']= 0 if idf <0 else idf    
    
      del token
      del ni
      del idf
      
    return diccionario
    


    ### Aquí termina tu código ###

***Test:***

In [ ]:
idf = calc_idf(dataset_bow)
idf

{'a': 0.47712125471966244,
 'la': 0.17609125905568124,
 'come': 0.17609125905568124,
 'y': 0.0,
 'despierta': 0.47712125471966244,
 'se': 0,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'El': 0.0,
 'maullar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'comida': 0.17609125905568124,
 'duerme': 0.47712125471966244,
 'ladrar': 0.7781512503836436,
 'después': 0.0,
 'ladra': 0.7781512503836436,
 'perro': 0.3010299956639812}

***Resultado esperado***: 

```python
{'El': 0.0, 
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}
```

Puede notar el bajo puntaje otorgado a las palabras que más se repiten! 😮

**Ejercicio 6 - *Calcular TF-IDF & concluir similitud de documentos.* (1 punto)**


Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [ ]:
def calc_tf_idf(tf, idf):
    ### Aquí inicia tu código ###

    return tf*idf

    ### Aquí termina tu código ### 

***Test.***

In [ ]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

a        la      come    y  despierta   se   empieza      gato  \
d0  0.000000  0.088046  0.088046  0.0   0.000000  0.0  0.000000  0.000000   
d1  0.477121  0.000000  0.000000  0.0   0.477121  0.0  0.477121  0.000000   
d2  0.000000  0.176091  0.176091  0.0   0.000000  0.0  0.000000  0.000000   
d3  0.000000  0.088046  0.088046  0.0   0.000000  0.0  0.000000  0.150515   
d4  0.477121  0.000000  0.000000  0.0   0.477121  0.0  0.477121  0.301030   
d5  0.000000  0.176091  0.176091  0.0   0.000000  0.0  0.000000  0.301030   

     El   maullar    maulla    comida    duerme    ladrar  después     ladra  \
d0  0.0  0.000000  0.000000  0.088046  0.238561  0.000000      0.0  0.000000   
d1  0.0  0.000000  0.000000  0.000000  0.000000  0.778151      0.0  0.000000   
d2  0.0  0.000000  0.000000  0.176091  0.000000  0.000000      0.0  0.778151   
d3  0.0  0.000000  0.000000  0.088046  0.238561  0.000000      0.0  0.000000   
d4  0.0  0.778151  0.000000  0.000000  0.000000  0.000000      0.0  0.000000   
d5  0.0  0.000000  0.778151  0.176091  0.000000  0.000000      0.0  0.000000   

       perro  
d0  0.150515  
d1  0.301030  
d2  0.301030  
d3  0.000000  
d4  0.000000  
d5  0.000000

``
``

***Resultado esperado***: 

|    |  El |    perro |  se |     come |       la |   comida |   y | después |   duerme | despierta |  empieza |        a |   ladrar |    ladra |     gato |  maullar |   maulla |
|----|----:|---------:|----:|---------:|---------:|---------:|----:|--------:|---------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|
| d0 | 0.0 | 0.150515 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d1 | 0.0 | 0.301030 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.778151 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d2 | 0.0 | 0.301030 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.778151 | 0.000000 | 0.000000 | 0.000000 |
| d3 | 0.0 | 0.000000 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.150515 | 0.000000 | 0.000000 |
| d4 | 0.0 | 0.000000 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.000000 | 0.000000 | 0.301030 | 0.778151 | 0.000000 |
| d5 | 0.0 | 0.000000 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.301030 | 0.000000 | 0.778151 |


``
``

Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica. Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos vectores. Concluya cuáles son los dos documentos más similares.

In [ ]:
def cosine_similarity(v1, v2):
  ### Aquí inicia tu código ###
  num=np.dot(v1,v2)
  den=np.linalg.norm(v1)*np.linalg.norm(v2)
  return num/den


  ### Aquí termina tu código ### 

similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity


In [ ]:
similarity_matrix

array([[1.        , 0.12032418, 0.3223436 , 0.77967014, 0.        ,
        0.1632828 ],
       [0.12032418, 1.        , 0.08686457, 0.        , 0.4952126 ,
        0.        ],
       [0.3223436 , 0.08686457, 1.        , 0.1632828 , 0.        ,
        0.11787732],
       [0.77967014, 0.        , 0.1632828 , 1.        , 0.12032418,
        0.3223436 ],
       [0.        , 0.4952126 , 0.        , 0.12032418, 1.        ,
        0.08686457],
       [0.1632828 , 0.        , 0.11787732, 0.3223436 , 0.08686457,
        1.        ]])

![gato](https://live.staticflickr.com/4652/38904147065_0b6c446945_b.jpg)

**Cualquier recomendación que nos quisieran dar para una futura tarea es bienvenid@!**